File: CS/T223817/CustomEdit.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Linq;
using System.Text;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Popup;
using System.Globalization;
using DevExpress.Data.Mask;

namespace T223817 {
    [UserRepositoryItem("RegisterCustomEdit")]
    public class RepositoryItemCustomEdit : RepositoryItemTextEdit {
        private static readonly object invalidCharacter = new object();
        static RepositoryItemCustomEdit() {
            RegisterCustomEdit();
        }

        public const string CustomEditName = "CustomEdit";
        public RepositoryItemCustomEdit() {
        }

        public override string EditorTypeName {
            get { return CustomEditName; }
        }

        public static void RegisterCustomEdit() {
            Image img = null;
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(CustomEditName, typeof(CustomEdit), typeof(RepositoryItemCustomEdit), typeof(TextEditViewInfo), new TextEditPainter(), true, img));
        }

        public override void Assign(RepositoryItem item) {
            RepositoryItemCustomEdit source = item as RepositoryItemCustomEdit;
            BeginUpdate();
            try {
                base.Assign(item);
                if (source == null) {
                    return;
                }
            }
            finally {
                EndUpdate();
            }
            Events.AddHandler(invalidCharacter, source.Events[invalidCharacter]);
        }

        public event InvalidCharacterEventHandler InvalidCharacter {
            add { this.Events.AddHandler(invalidCharacter, value); }
            remove { this.Events.RemoveHandler(invalidCharacter, value); }
        }

        protected internal virtual void RaiseInvalidCharacter(InvalidCharacterEventArgs e) {
            InvalidCharacterEventHandler handler = (InvalidCharacterEventHandler)this.Events[invalidCharacter];
            if (handler != null) {
                handler(GetEventSender(), e);
            }
        }
    }

    public delegate void InvalidCharacterEventHandler(object sender, InvalidCharacterEventArgs e);
    public class InvalidCharacterEventArgs : EventArgs {
        public string InsertionString { get; private set; }
        public bool Handled { get; set; }
        public InvalidCharacterEventArgs(string insertionString) {
            InsertionString = insertionString;
        }
    }

    [ToolboxItem(true)]
    public class CustomEdit : TextEdit {
        static CustomEdit() {
            RepositoryItemCustomEdit.RegisterCustomEdit();
        }

        public CustomEdit() {
        }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemCustomEdit Properties {
            get { return base.Properties as RepositoryItemCustomEdit; }
        }

        public override string EditorTypeName {
            get { return RepositoryItemCustomEdit.CustomEditName; }
        }

        private NumericMaskManager fManager;
        protected override MaskManager CreateMaskManager(DevExpress.XtraEditors.Mask.MaskProperties mask) {
            fManager = (NumericMaskManager)base.CreateMaskManager(mask);
            return fManager;
        }

        protected override void OnKeyPress(System.Windows.Forms.KeyPressEventArgs e) {
            base.OnKeyPress(e);
            string insertion = KeyCharToInsertableString(e.KeyChar);
            if (insertion == null) {
                return;
            }
            if (fManager.Insert(insertion)) {
                fManager.Backspace();
                return;
            }
            ProcessInvalidCharacter(insertion);
        }

        private void ProcessInvalidCharacter(string insertion) {
            InvalidCharacterEventArgs args = new InvalidCharacterEventArgs(insertion);
            this.Properties.RaiseInvalidCharacter(args);
            if (args.Handled) {
                return;
            }
            this.ToolTipController.ShowBeak = true;
            this.ToolTipController.ShowHint("An editor can accept only numbers or maximum length achieved", this.PointToScreen(Point.Empty));
        }

        private string KeyCharToInsertableString(char keyChar) {
            if (!char.IsControl(keyChar)) {
                return keyChar.ToString(CultureInfo.InvariantCulture);
            }
            if (keyChar == '\r' && this.MaskBox.Multiline && this.MaskBox.AcceptsReturn) {
                return "\r\n";
            }
            if (keyChar == '\t' && this.MaskBox.Multiline && this.MaskBox.AcceptsTab) {
                return "\t";
            }
            return null;
        }
    }
}

File: VB/T223817/CustomEdit.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Popup
Imports System.Globalization
Imports DevExpress.Data.Mask

Namespace T223817
    <UserRepositoryItem("RegisterCustomEdit")> _
    Public Class RepositoryItemCustomEdit
        Inherits RepositoryItemTextEdit

        Private Shared ReadOnly _invalidCharacter As New Object()
        Shared Sub New()
            RegisterCustomEdit()
        End Sub

        Public Const CustomEditName As String = "CustomEdit"
        Public Sub New()
        End Sub

        Public Overrides ReadOnly Property EditorTypeName() As String
            Get
                Return CustomEditName
            End Get
        End Property

        Public Shared Sub RegisterCustomEdit()
            Dim img As Image = Nothing
            EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(CustomEditName, GetType(CustomEdit), GetType(RepositoryItemCustomEdit), GetType(TextEditViewInfo), New TextEditPainter(), True, img))
        End Sub

        Public Overrides Sub Assign(ByVal item As RepositoryItem)
            Dim source As RepositoryItemCustomEdit = TryCast(item, RepositoryItemCustomEdit)
            BeginUpdate()
            Try
                MyBase.Assign(item)
                If source Is Nothing Then
                    Return
                End If
            Finally
                EndUpdate()
            End Try
            Events.AddHandler(_invalidCharacter, source.Events(_invalidCharacter))
        End Sub

        Public Custom Event InvalidCharacter As InvalidCharacterEventHandler
            AddHandler(ByVal value As InvalidCharacterEventHandler)
                Me.Events.AddHandler(_invalidCharacter, value)
            End AddHandler
            RemoveHandler(ByVal value As InvalidCharacterEventHandler)
                Me.Events.RemoveHandler(_invalidCharacter, value)
            End RemoveHandler
            RaiseEvent(ByVal sender As Object, ByVal e As InvalidCharacterEventArgs)
            End RaiseEvent
        End Event

        Protected Friend Overridable Sub RaiseInvalidCharacter(ByVal e As InvalidCharacterEventArgs)
            Dim handler As InvalidCharacterEventHandler = CType(Me.Events(_invalidCharacter), InvalidCharacterEventHandler)
            If handler IsNot Nothing Then
                handler(GetEventSender(), e)
            End If
        End Sub
    End Class

    Public Delegate Sub InvalidCharacterEventHandler(ByVal sender As Object, ByVal e As InvalidCharacterEventArgs)
    Public Class InvalidCharacterEventArgs
        Inherits EventArgs

        Private privateInsertionString As String
        Public Property InsertionString() As String
            Get
                Return privateInsertionString
            End Get
            Private Set(ByVal value As String)
                privateInsertionString = value
            End Set
        End Property
        Public Property Handled() As Boolean
        Public Sub New(ByVal insertionString As String)
            Me.InsertionString = insertionString
        End Sub
    End Class

    <ToolboxItem(True)> _
    Public Class CustomEdit
        Inherits TextEdit

        Shared Sub New()
            RepositoryItemCustomEdit.RegisterCustomEdit()
        End Sub

        Public Sub New()
        End Sub

        <DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
        Public Shadows ReadOnly Property Properties() As RepositoryItemCustomEdit
            Get
                Return TryCast(MyBase.Properties, RepositoryItemCustomEdit)
            End Get
        End Property

        Public Overrides ReadOnly Property EditorTypeName() As String
            Get
                Return RepositoryItemCustomEdit.CustomEditName
            End Get
        End Property

        Private fManager As NumericMaskManager
        Protected Overrides Function CreateMaskManager(ByVal mask As DevExpress.XtraEditors.Mask.MaskProperties) As MaskManager
            fManager = CType(MyBase.CreateMaskManager(mask), NumericMaskManager)
            Return fManager
        End Function

        Protected Overrides Sub OnKeyPress(ByVal e As System.Windows.Forms.KeyPressEventArgs)
            MyBase.OnKeyPress(e)
            Dim insertion As String = KeyCharToInsertableString(e.KeyChar)
            If insertion Is Nothing Then
                Return
            End If
            If fManager.Insert(insertion) Then
                fManager.Backspace()
                Return
            End If
            ProcessInvalidCharacter(insertion)
        End Sub

        Private Sub ProcessInvalidCharacter(ByVal insertion As String)
            Dim args As New InvalidCharacterEventArgs(insertion)
            Me.Properties.RaiseInvalidCharacter(args)
            If args.Handled Then
                Return
            End If
            Me.ToolTipController.ShowBeak = True
            Me.ToolTipController.ShowHint("An editor can accept only numbers or maximum length achieved", Me.PointToScreen(Point.Empty))
        End Sub

        Private Function KeyCharToInsertableString(ByVal keyChar As Char) As String
            If Not Char.IsControl(keyChar) Then
                Return keyChar.ToString(CultureInfo.InvariantCulture)
            End If
            If keyChar = ControlChars.Cr AndAlso Me.MaskBox.Multiline AndAlso Me.MaskBox.AcceptsReturn Then
                Return ControlChars.CrLf
            End If
            If keyChar = ControlChars.Tab AndAlso Me.MaskBox.Multiline AndAlso Me.MaskBox.AcceptsTab Then
                Return ControlChars.Tab
            End If
            Return Nothing
        End Function
    End Class
End Namespace
